### PR TITLE
Allow to expect (adopted)

### LINKED ISSUE #15059: Swap most uses of allow in lints to expect
## What problem does this solve or what need does it fill?

With the advent of [Rust 1.81](https://blog.rust-lang.org/2024/09/05/Rust-1.81.0.html), we can now distinguish between lints that we haven't gotten around to fixing and those where the lint is wrong.

## What solution would you like?

Most localized lints in Bevy should be `expect`: the only exception that comes to mind is `missing_docs`.

Leave the `Cargo.toml` lints set as allow, as funny as it would be to set a crate-level `expect(type_complexity)`.

### PR DESCRIPTION
**Pull request by richchurcher**
# Objective

> Rust 1.81 released the #[expect(...)] attribute, which works like #[allow(...)] but throws a warning if the lint isn't raised. This is preferred to #[allow(...)] because it tells us when it can be removed.

- Adopts the parts of #15118 that are complete, and updates the branch so it can be merged.
- There were a few conflicts, let me know if I misjudged any of 'em.

Alice's [recommendation](https://github.com/bevyengine/bevy/issues/15059#issuecomment-2349263900) seems well-taken, let's do this crate by crate now that @BD103 has done the lion's share of this!

(Relates to, but doesn't yet completely finish #15059.)

Crates this _doesn't_ cover:

- bevy_input
- bevy_gilrs
- bevy_window
- bevy_winit
- bevy_state
- bevy_render
- bevy_picking
- bevy_core_pipeline
- bevy_sprite
- bevy_text
- bevy_pbr
- bevy_ui
- bevy_gltf
- bevy_gizmos
- bevy_dev_tools
- bevy_internal
- bevy_dylib

### CODE REVIEW

**Comment by BD103** [comment]
Thanks for picking this up! Ping me if you have any questions about how I approached this problem :)

File: crates/bevy_app/src/plugin_group.rs
Comment (alice-i-cecile) [inline crates/bevy_app/src/plugin_group.rs:286]: We should rename `PluginGroupBuilder::add` to `queue`.

File: crates/bevy_log/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_log/src/lib.rs:72]: This is so much more useful, wow.

File: crates/bevy_reflect/derive/src/derive_data.rs
Comment (alice-i-cecile) [inline crates/bevy_reflect/derive/src/derive_data.rs:126]: @MrGVSV can you help out with this one?

**Review by alice-i-cecile — APPROVED** [review]
LGTM. I'd prefer to resolve that one TODO before merging but I won't block on it.

File: Cargo.toml
Comment (alice-i-cecile) [inline Cargo.toml:43]: This needs to be removed for now to get the PR mergeable.

File: crates/bevy_asset/src/processor/mod.rs
Comment (chompaa) [inline crates/bevy_asset/src/processor/mod.rs:564]:
```suggestion
            reason = "This function is only used when the `multi_threaded` feature is enabled, and when not on WASM."
```

File: crates/bevy_transform/src/systems.rs
Comment (chompaa) [inline crates/bevy_transform/src/systems.rs:82]:
```suggestion
                #[expect(unsafe_code, reason = "`propagate_recursive()` is unsafe due to its use of `Query::get_unchecked()`.")]
```

**Review by chompaa — CHANGES_REQUESTED** [review]
Two minor nits, otherwise LGTM.

File: crates/bevy_reflect/derive/src/derive_data.rs
Comment (richchurcher) [reply crates/bevy_reflect/derive/src/derive_data.rs:126]: Oh, I think they said in the previous PR that it could be removed, but I wasn't sure so of the implications so left the field there for the meantime.

File: Cargo.toml
Comment (richchurcher) [reply Cargo.toml:43]: Ah, got it. Will do.

File: crates/bevy_reflect/derive/src/derive_data.rs
Comment (richchurcher) [reply crates/bevy_reflect/derive/src/derive_data.rs:126]: https://github.com/bevyengine/bevy/pull/15118#discussion_r1750326458

Comment (alice-i-cecile) [reply crates/bevy_reflect/derive/src/derive_data.rs:126]: Please remove it now then :)

**Comment by alice-i-cecile** [comment]
@BenjaminBrienen could I get your review here? :) Should all be straightforward; I'd just like a second set of eyes.

**Comment by richchurcher** [comment]
I should remember not to rebase when there's a rapidly-moving `main` :confused: force of habit.

**Comment by alice-i-cecile** [comment]
Yeah, I generally find merging a lot less painful when working with Bevy.

File: crates/bevy_asset/src/io/embedded/mod.rs
Comment (BenjaminBrienen) [inline crates/bevy_asset/src/io/embedded/mod.rs:39]: Can we improve the API by instead changing the function signature on this feature?

File: crates/bevy_asset/src/io/embedded/mod.rs
Comment (BenjaminBrienen) [inline crates/bevy_asset/src/io/embedded/mod.rs:75]: This comment is now out of date and can be removed

File: crates/bevy_asset/src/processor/mod.rs
Comment (BenjaminBrienen) [inline crates/bevy_asset/src/processor/mod.rs:80]: This only affects 8 total documentation links. Looking at other documentation comments, it appears that we prefer to fully qualify the link within the documentation comment rather than doing this.

File: crates/bevy_asset/src/io/source.rs
Comment (BenjaminBrienen) [inline crates/bevy_asset/src/io/source.rs:520]: Can we improve the API by instead changing the function signature on this feature/arch/os?

File: crates/bevy_asset/src/loader.rs
Comment (BenjaminBrienen) [inline crates/bevy_asset/src/loader.rs:260]: Do we not prefer to use `Self` in impl blocks?

File: crates/bevy_asset/src/io/embedded/mod.rs
Comment (richchurcher) [reply crates/bevy_asset/src/io/embedded/mod.rs:39]: It seems likely, but should we create an issue rather than folding an API change into this PR?

File: crates/bevy_asset/src/loader.rs
Comment (BenjaminBrienen) [reply crates/bevy_asset/src/loader.rs:260]: You could also mention that the "large error variant" is only 10% bigger than the okay variant.

File: crates/bevy_asset/src/processor/mod.rs
Comment (BenjaminBrienen) [inline crates/bevy_asset/src/processor/mod.rs:566]: Would it make sense to instead make intialize() feature-gated?

File: crates/bevy_asset/src/processor/mod.rs
Comment (alice-i-cecile) [reply crates/bevy_asset/src/processor/mod.rs:80]: Yep; I'd prefer to fix this but won't block on it being done in this PR.

File: crates/bevy_asset/src/io/embedded/mod.rs
Comment (BenjaminBrienen) [reply crates/bevy_asset/src/io/embedded/mod.rs:39]: Yeah, would you like to do that for each of those?

File: crates/bevy_asset/src/loader.rs
Comment (alice-i-cecile) [reply crates/bevy_asset/src/loader.rs:260]: Generally we don't have strong feelings. I prefer Self here though.

File: crates/bevy_asset/src/io/embedded/mod.rs
Comment (alice-i-cecile) [reply crates/bevy_asset/src/io/embedded/mod.rs:58]: Doing this will be harder to maintain and will require platform specific code downstream.

File: crates/bevy_asset/src/io/source.rs
Comment (alice-i-cecile) [reply crates/bevy_asset/src/io/source.rs:520]: Same comment as above: we shouldn't do this.

File: crates/bevy_asset/src/io/source.rs
Comment (BenjaminBrienen) [inline crates/bevy_asset/src/io/source.rs:14]: Same as https://github.com/bevyengine/bevy/pull/15301/files#r1767672248

**Comment by richchurcher** [comment]
Troubleshooting crate-level `expect`, they seem to all be failing :thinking:

File: crates/bevy_asset/src/io/embedded/mod.rs
Comment (richchurcher) [reply crates/bevy_asset/src/io/embedded/mod.rs:39]: No, I think I have my hands full already :sweat_smile:

File: crates/bevy_asset/src/reflect.rs
Comment (BenjaminBrienen) [inline crates/bevy_asset/src/reflect.rs:60]: I wonder why rustfmt doesn't remove the unnecessary parentheses here.

File: crates/bevy_asset/src/server/info.rs
Comment (BenjaminBrienen) [inline crates/bevy_asset/src/server/info.rs:118]: An issue should be created to reduce the complexity of this function.

Comment (BenjaminBrienen) [reply crates/bevy_asset/src/server/info.rs:118]: pub(crate) fn create_loading_handle_untyped(
        &mut self,
        type_id: TypeId,
        type_name: &'static str,
    ) -> UntypedHandle {
        unwrap_with_context(
            (|| -> Result<UntypedHandle, GetOrCreateHandleInternalError> {
                let provider = &self
                    .handle_providers
                    .get(&type_id)
                    .ok_or(MissingHandleProviderError(type_id))?;
                Ok(Self::create_handle_internal(
                    provider,
                    &mut self.infos,
                    &mut self.living_labeled_assets,
                    self.watching_for_changes,
                    None,
                    None,
                    true,
                ))
            })(),
            type_name,
        )
        .unwrap()
    }

    fn create_handle_internal(
        provider: &&AssetHandleProvider,
        infos: &mut HashMap<UntypedAssetId, AssetInfo>,
        living_labeled_assets: &mut HashMap<AssetPath<'static>, HashSet<Box<str>>>,
        watching_for_changes: bool,
        path: Option<AssetPath<'static>>,
        meta_transform: Option<MetaTransform>,
        loading: bool,
    ) -> UntypedHandle {
        if watching_for_changes {
            if let Some(path) = &path {
                let mut without_label = path.to_owned();
                if let Some(label) = without_label.take_label() {
                    let labels = living_labeled_assets.entry(without_label).or_default();
                    labels.insert(label.as_ref().into());
                }
            }
        }
        let handle = provider.reserve_handle_internal(true, path.clone(), meta_transform);
        let mut info = AssetInfo::new(Arc::downgrade(&handle), path);
        if loading {
            info.load_state = LoadState::Loading;
            info.dep_load_state = DependencyLoadState::Loading;
            info.rec_dep_load_state = RecursiveDependencyLoadState::Loading;
        }
        infos.insert(handle.id, info);
        UntypedHandle::Strong(handle)
    }

...

            Entry::Vacant(entry) => {
                let should_load = match loading_mode {
                    HandleLoadingMode::NotLoading => false,
                    HandleLoadingMode::Request | HandleLoadingMode::Force => true,
                };
                let provider = &self
                    .handle_providers
                    .get(&type_id)
                    .ok_or(MissingHandleProviderError(type_id))?;
                let handle = Self::create_handle_internal(
                    provider,
                    &mut self.infos,
                    &mut self.living_labeled_assets,
                    self.watching_for_changes,
                    Some(path),
                    meta_transform,
                    should_load,
                );
                entry.insert(handle.id());
                Ok((handle, should_load))
            }

Comment (BenjaminBrienen) [reply crates/bevy_asset/src/server/info.rs:118]: The lint doesn't trigger with this quick change

File: crates/bevy_asset/src/server/mod.rs
Comment (BenjaminBrienen) [inline crates/bevy_asset/src/server/mod.rs:36]: Same as https://github.com/bevyengine/bevy/pull/15301/files#r1767672248

File: crates/bevy_encase_derive/src/lib.rs
Comment (BenjaminBrienen) [inline crates/bevy_encase_derive/src/lib.rs:1]: Indicate the open issue (or solve 😄)

File: crates/bevy_mikktspace/src/lib.rs
Comment (BenjaminBrienen) [inline crates/bevy_mikktspace/src/lib.rs:7]: This is mega-pedantic so feel free to ignore:
This reason message has a grammatical error called a comma splice, where two independent clauses are joined by a comma.
2 possible fixes:
 - "Not all docs are written yet. See #3492."
 - "Not all docs are written yet; see #3492."

File: crates/bevy_asset/src/io/embedded/mod.rs
Comment (BenjaminBrienen) [reply crates/bevy_asset/src/io/embedded/mod.rs:39]: Ok, I can create them after this is merged. I'll wait because there might be more and it'll be easier to do all at once.

File: crates/bevy_picking/src/backend.rs
Comment (BenjaminBrienen) [inline crates/bevy_picking/src/backend.rs:87]: missing reason

File: crates/bevy_picking/src/backend.rs
Comment (BenjaminBrienen) [inline crates/bevy_picking/src/backend.rs:115]: missing reason

File: crates/bevy_reflect/src/lib.rs
Comment (BenjaminBrienen) [inline crates/bevy_reflect/src/lib.rs:1]: same as above

File: crates/bevy_render/src/lib.rs
Comment (BenjaminBrienen) [inline crates/bevy_render/src/lib.rs:1]: same as above

File: crates/bevy_mikktspace/src/lib.rs
Comment (richchurcher) [reply crates/bevy_mikktspace/src/lib.rs]: I take your point, but given it doesn't interfere with comprehension in this case I'll turn a blind eye to grammar :slightly_smiling_face:

File: crates/bevy_asset/src/server/info.rs
Comment (richchurcher) [reply crates/bevy_asset/src/server/info.rs:118]: Looks like Alice has you covered! #15322

File: crates/bevy_asset/src/processor/mod.rs
Comment (richchurcher) [reply crates/bevy_asset/src/processor/mod.rs:80]: @BenjaminBrienen chuck it in an issue, I'll avoid making that change here :slightly_smiling_face:

Comment (alice-i-cecile) [reply crates/bevy_asset/src/processor/mod.rs:80]: Done.

File: crates/bevy_picking/src/backend.rs
Comment (alice-i-cecile) [reply crates/bevy_picking/src/backend.rs:87]: IMO we should just add trivial docs to these methods.

**Comment by richchurcher** [comment]
I dare you to fail, CI.

**Comment by richchurcher** [comment]
I should not have dared :laughing:

**Review by james7132 — APPROVED** [review]

**Pull request merged**
